#!/usr/bin/env Rscript

# ------------------------------------------------------------------------------
# Copyright (c) 2021 CRUK Cambridge Institute - Bioinformatics Core

# Licensed under the MIT license (http://opensource.org/licenses/MIT).
# This file may not be copied, modified, or distributed except according
# to those terms.
# ------------------------------------------------------------------------------

# Extracts annotations from VCF file generated by Ensembl Variant Effect
# Predictor into a tabular format.

library(optparse)

option_list <- list(
  make_option(c("--vcf"), dest = "vep_vcf_file",
              help = "VCF file containing Ensembl VEP annotations"),
  make_option(c("--output"), dest = "output_file",
              help = "Tab-delimited output file")
)

option_parser <- OptionParser(usage = "usage: %prog [options]", option_list = option_list, add_help_option = TRUE)
opt <- parse_args(option_parser)

vep_vcf_file <- opt$vep_vcf_file
output_file <- opt$output_file

if (is.null(vep_vcf_file)) stop("Ensembl VEP output VCF file must be specified")
if (is.null(output_file)) stop("Output file must be specified")

suppressPackageStartupMessages(library(tidyverse))

vep_annotation_types <- tibble(name = readLines(vep_vcf_file)) %>%
  filter(str_detect(name, "INFO=<ID=CSQ")) %>%
  mutate(name = str_remove(name, "^.*Format: ")) %>%
  mutate(name = str_remove(name, "\">$")) %>%
  separate_rows(name, sep = "\\|") %>%
  transmute(index = row_number(), name)

variants <- read_tsv(
  vep_vcf_file,
  comment = "##",
  col_types = cols(
    `#CHROM` = "f",
    POS = "i",
    ID = "c",
    REF = "c",
    ALT = "c",
    QUAL = "c",
    FILTER = "c",
    INFO = "c",
    .default = "c"
  )
)

vep_annotations <- variants %>%
  select(Chromosome = `#CHROM`, Position = POS, Ref = REF, Alt = ALT, value = INFO) %>%
  mutate(value = str_remove(value, "^CSQ=")) %>%
  separate_rows(value, sep = ",") %>%
  mutate(variant_annotation_number = row_number()) %>%
  separate_rows(value, sep = "\\|") %>%
  group_by(variant_annotation_number) %>%
  mutate(index = row_number()) %>%
  ungroup() %>%
  left_join(vep_annotation_types, by = "index") %>%
  select(-index) %>%
  pivot_wider(names_from = "name", values_from = "value") %>%
  select(
    Chromosome, Position, Ref, Alt,
    `Variant type` = VARIANT_CLASS,
    Gene,
    `Gene symbol` = SYMBOL,
    Feature,
    `Feature type` = Feature_type,
    Exon = EXON,
    Intron = INTRON,
    Consequence,
    Impact = IMPACT,
    `Existing variation` = Existing_variation,
    `Clinical significance` = CLIN_SIG,
    PubMed = PUBMED,
    SIFT,
    any_of("PolyPhen"), # note that PolyPhen is disabled for some genomes so this column may not exist
    Codons,
    `cDNA effect` = HGVSc,
    `Protein effect` = HGVSp,
    `1000 Genomes frequency` = AF,
    `gnomAD frequency` = gnomAD_AF
  )

# reformats the HGVS protein effect from Ensembl VEP replacing 3-letter amino
# acid codes with 1-letter equivalents, replacing = with the amino acid code for
# synonymous/silent changes and replacing Ter with * for termination codons
reformat_protein_effect <- function(protein_effect) {
  protein_effect %>%
    str_remove("^[A-Z0-9\\.]+:") %>%
    str_replace_all("Ala", "A") %>%
    str_replace_all("Cys", "C") %>%
    str_replace_all("Asp", "D") %>%
    str_replace_all("Glu", "E") %>%
    str_replace_all("Phe", "F") %>%
    str_replace_all("Gly", "G") %>%
    str_replace_all("His", "H") %>%
    str_replace_all("Ile", "I") %>%
    str_replace_all("Lys", "K") %>%
    str_replace_all("Leu", "L") %>%
    str_replace_all("Met", "M") %>%
    str_replace_all("Asn", "N") %>%
    str_replace_all("Pro", "P") %>%
    str_replace_all("Gln", "Q") %>%
    str_replace_all("Arg", "R") %>%
    str_replace_all("Ser", "S") %>%
    str_replace_all("Thr", "T") %>%
    str_replace_all("Val", "V") %>%
    str_replace_all("Trp", "W") %>%
    str_replace_all("Tyr", "Y") %>%
    str_replace_all("Ter", "*") %>%
    str_replace("(?<=^p\\.([A-Z])[0-9]{1,10})=$", "\\1")
}

vep_annotations <- vep_annotations %>%
  mutate(across(!Chromosome:Alt, ~ str_replace_all(.x, "&", ", "))) %>%
  mutate(Consequence = str_remove_all(Consequence, "_variant")) %>%
  mutate(Consequence = str_replace(Consequence, "missense", "nonsynonymous")) %>%
  mutate(Impact = str_to_lower(Impact)) %>%
  mutate(across(c(Exon, Intron), ~ str_replace(.x, "/", " of "))) %>%
  mutate(`cDNA effect` = str_remove(`cDNA effect`, "^[A-Z0-9\\.]+:")) %>%
  mutate(`Protein effect` = reformat_protein_effect(`Protein effect`))

write_tsv(vep_annotations, output_file)

